### PR TITLE
[WIP] Handle Peer Banning

### PR DESCRIPTION
**Pull request by Latrasis**
Based on #56. Implement minimal peer banning.

* [x] Error Handling from `NetAdapter`
* [x] Setup Testing

### CODE REVIEW

File: p2p/src/peer.rs [hunk 1]
```diff
@@ -95,12 +95,17 @@ impl Peer {
...
 					Ok(())
-				}
+				},
 				Err(Error::Serialization(e)) => {
 					*state = State::Banned;
 					info!("Client {} corrupted, ban.", addr);
 					Err(Error::Serialization(e))
-				}
+				},
+				Err(Error::Invalid) => {
```
Comment (Latrasis) [inline p2p/src/peer.rs]: Still not sure if this is the behavior you are looking for.

File: grin/src/adapters.rs [hunk 2]
```diff
@@ -85,14 +91,15 @@ impl NetAdapter for NetToChainAdapter {
 					      bh.hash(),
 					      bh.height,
 					      s);
+					return Err(p2p::Error::Invalid);
```
Comment (ignopeverell) [inline grin/src/adapters.rs]: This gets returned when a peer sent a valid header but at the wrong time (too early/late with respect to our current chain state). We should not penalize the peer for that.

File: grin/src/adapters.rs [hunk 3]
```diff
@@ -85,14 +91,15 @@ impl NetAdapter for NetToChainAdapter {
...
 				}
 				Err(chain::Error::StoreErr(e)) => {
 					error!("Store error processing block header {}: {:?}", bh.hash(), e);
-					return;
+					return Err(io::Error::new(io::ErrorKind::InvalidInput, "Invalid Header Format").into());
 				}
 				Err(e) => {
 					info!("Invalid block header {}: {:?}.", bh.hash(), e);
-					// TODO penalize peer somehow
+					return Err(p2p::Error::Invalid);
```
Comment (ignopeverell) [inline grin/src/adapters.rs]: Yes, this is good.

File: p2p/src/peer.rs (hunk 1 above)
Comment (ignopeverell) [inline p2p/src/peer.rs]: Yes, if the data was confirmed invalid, banning is appropriate.

**Comment by Latrasis** [comment]
Tbh, I would withhold testing for this.

**Comment by ignopeverell** [comment]
This lends itself much more to integration tests. I would just have a peer connect to another, send a bad message (like a block with a bad PoW) and ensure it got disconnected and banned.

**Comment by ignopeverell** [comment]
Is there some way I can be more helpful with this? Want to setup some time to chat on gitter or ping me there to discuss?

**Comment by Latrasis** [comment]
@ignopeverell: Apologies for the delay, I'm happy to discuss. (Ping'ed on gitter)

**Comment by Latrasis** [comment]
@ignopeverell: Done at 942af4b. Read for review.

**Comment by ignopeverell** [comment]
PR looks good! Can you rebase master on your branch? Looks like there are some conflicts.

**Comment by Latrasis** [comment]
Sure thing

**Comment by ignopeverell** [comment]
Pinging, just in case you haven't noticed the test failures.

**Comment by Latrasis** [comment]
Saw that, some issues with windows. https://github.com/mimblewimble/cuckoo-miner/issues/1. I'll try fixing on a different machine sometime today.

**Comment by Latrasis** [comment]
Could we try with a sequential run?

**Comment by ignopeverell** [comment]
I retried the p2p module by itself and it still failed. Thinking it's 2 tests within p2p? You can always set RUST_TEST_THREADS=1 in the travis config in your branch.

**Comment by Latrasis** [comment]
Hmm, for me, the `peer_banning` test passes on its own, but even with single thread it fails with `peer_handshake`.

**Comment by ignopeverell** [comment]
I'd configure everything to work on different ports across tests. I've noticed Rust is a little lazy in closing those sockets.

**Comment by Latrasis** [comment]
@ignopeverell: Getting a segfault: https://travis-ci.org/ignopeverell/grin/jobs/260839727

**Comment by ignopeverell** [comment]
Going to try the build a few times to see if it helps. Otherwise we may need #85 to get a reliable build.

**Comment by ignopeverell** [comment]
I merged a change that touched a lot of files, can I ask you to rebase? Then tests should pass.

**Comment by ignopeverell** [comment]
CI is still failing with a port already used error. I'll have to take a closer look at what's happening.

**Pull request closed without merging**